comp/core/secrets/secretsimpl/secrets.go

```diff
@@ -196,20 +205,44 @@ func isEnc(str string) (bool, string) {
 	return false, ""
 }
 
-// Resolve replaces all encrypted secrets in data by executing "secret_backend_command" once if all secrets aren't
-// present in the cache.
-func (r *secretResolver) Resolve(data []byte, origin string) ([]byte, error) {
-	return r.resolve(data, origin, nil)
+func (r *secretResolver) startRefreshRoutine() {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	// only start this routine once, and only if secret_refresh_interval is non-zero
+	if r.refreshInterval > 0 && r.lastRefresh.IsZero() {
+		r.lastRefresh = time.Now()
+		go func() {
+			for {
+				var since time.Duration
+				func() {
+					r.lock.Lock()
+					defer r.lock.Unlock()
+					since = time.Since(r.lastRefresh)
+				}()
+				if since.Seconds() > float64(r.refreshInterval) {
+					if err := r.Refresh(); err != nil {
+						log.Info(err)
+					}
+					since = 0
+				}
+				time.Sleep(time.Duration(r.refreshInterval) - since)
+			}
+		}()
+	}
 }
 
-// ResolveWithCallback resolves the secrets in the given yaml data calling the callback with the YAML path of
-// the secret handle and its value
-func (r *secretResolver) ResolveWithCallback(data []byte, origin string, cb secrets.ResolveCallback) error {
-	_, err := r.resolve(data, origin, cb)
-	return err
+// Subscribe adds this callback to the list that get notified when secrets are resolved or refreshed
+func (r *secretResolver) Subscribe(cb secrets.ResolveCallback) {
+	r.startRefreshRoutine()
+	r.subscriptions = append(r.subscriptions, cb)
```

Comment:
I think there is a race here, you write to `r.subscriptions` without locking, it can be read concurrently ?

Reply:
Nice catch 🎉

Reply:
Sure that makes sense, moved the lock from `startRefreshRoutine` into here.